Add input cleaner

djdv opened the pull request:
So that input such as `http://localhost:8080/ipfs/QmcbLLZJa9oXnYTZWfS4T16n8B2mHFZkAzkKBNdGVzgG8g`, `https://ipfs.io/ipfs/QmcbLLZJa9oXnYTZWfS4T16n8B2mHFZkAzkKBNdGVzgG8g`, etc. may be passed in and turned into proper references (`/ipfs/QmcbLLZJa9oXnYTZWfS4T16n8B2mHFZkAzkKBNdGVzgG8g`)
If nothing is found the input string is returned so that it may pass or fail by `path.ParsePath` later.

Blocked by: https://github.com/ipfs/ipget/pull/50
submitting now so I don't forget.

djdv commented:
Specifically I originally wanted this to do this:
https://github.com/djdv/go-ipfs/commit/0d250ccd097c949f60930eeefb0cd197e20dac24

djdv commented:
Pinging @Stebalien for opinions on this.
I feel like it's more likely that people are going to have url's on their clipboard and in their scripts than bare canonical paths. They should probably be supported as valid input.

Stebalien commented:
I'm really weary of stuff like this. What if we were to match URLs only? That is, check:

1. Does it start with `/`.
2. Does it start with a CID.
3. Does it start with `https?://foo.com/(ipfs|ipns)/...`?

djdv commented:
I changed this so that we first try to use `go-path`'s parser which should handle 1 & 2, and return early. 
If that fails, we check if the path looks like a URL and pass the found string to `go-path` again.

Regex changed ``^.*(/ip(f|n)s/.+)$`` (group anything in the string that looks like an ipfs/ipns path)
-> 
`^https?.*(/(ipfs|ipns|ipld)/.+)$` (specifically try to clamp to HTTP gateways, added IPLD)

Added a simple test as well.
(Ignore the force push spam.)

djdv commented:
This may be a better variant of the regex string.
`^https?://.*(/(?:ipfs|ipns|ipld)/.{46}(?:/.+)?$)` 
match HTTP, check for {/namespace/}, {fixed length for cid} {optional subpath}

If that's too specific
`^https?.*(/(ipfs|ipns|ipld)/.+)$` should still probably be changed to
`^https?.*(/(?:ipfs|ipns|ipld)/.+)$` (don't capture the namespace, we have no need for that)

Stebalien commented on main.go:
```diff
@@ -132,3 +131,17 @@ func movePostfixOptions(args []string) []string {
+		return ipfsPath, nil
+	}
+	// allow HTTP gateway URLs as input as well
+	matches := regexp.MustCompile(`^https?://.*(/(?:ipfs|ipns|ipld)/.+)$`).FindStringSubmatch(path)
```
`*` is greedy. This'll parse `http://bla.com/ipfs/QmA/ipfs/foobar` as `/ipfs/foobar`. What if we just parsed the input as a URL and took the path? IMO, it's reasonable to assume that gateways are run on their own origin (for security reasons).

djdv commented:
>What if we just parsed the input as a URL and took the path?

Much better!

I also added the handling for these https://github.com/ipfs-shipyard/ipfs-companion/pull/533
`ipfs://QmcbLLZJa9oXnYTZWfS4T16n8B2mHFZkAzkKBNdGVzgG8g`
and it seems fine.
But I'm not sure if this is wanted or not. (revertible)

Stebalien commented on main.go:
```diff
@@ -132,3 +134,21 @@ func movePostfixOptions(args []string) []string {
+
+	switch proto := u.Scheme; proto {
+	case "ipfs", "ipld", "ipns":
+		return ipath.ParsePath(gopath.Join(fmt.Sprintf("/%s/", proto), u.Host, u.Path))
```
Nit: We can just write `gopath.Join("/", proto, u.Host, u.Path)`.

Stebalien commented on main.go:
```diff
@@ -132,3 +134,21 @@ func movePostfixOptions(args []string) []string {
+	switch proto := u.Scheme; proto {
+	case "ipfs", "ipld", "ipns":
+		return ipath.ParsePath(gopath.Join(fmt.Sprintf("/%s/", proto), u.Host, u.Path))
+	default:
```
I'd be conservative for now.

`case "https", "http":`
